text/0000-fields-in-traits.md

```diff
@@ -0,0 +1,800 @@
+   ...
+   field: lvalue,
+   ...
+}
+```
+
+then `lvalue` must be *disjoint* (defined below) from the lvalues
+found in all other field mappings within that impl. Furthermore,
+`lvalue` must be disjoint from the lvalues mapped in supertrait impls.
+
+Note that if a trait has two unrelated supertraits, the fields of those
+supertraits do not have to be disjoint. This can be used to form arbitrary
+disjointness relationships. For example, imagine that we have some type `Foo`
+that implements the trait `C` shown below:
+
+```rust
+trait A {
+    f: u32,
+    g: u32, // must be disjoint from f
+}
+
+trait B {
+    h: u32, // can overlap with f and g
+}
+
+trait C: A + B {
+    i: u32, // must be disjoint from f, g, and h
+}
+```
+
+As the comments suggest, `Foo` could map `A::f` and `B::h` to the same
+lvalue if it wanted to, but it must map `A::f`, `A::g`, and `C::i` to
+mutually disjoint lvalues.
+
+#### Definition of "disjoint"
+
+Two lvalues are considered "disjoint" if the borrow checker would
+allow `&mut` borrows of them simultaneously. This is typically true
+because they are based in different fields of the same struct.
+
+### Fully qualified field access
+
+Generally speaking, the lookup for fields will follow the same rules
+as methods: if the type itself defines an "inherent" field, that field
+is used, but otherwise we search for in-scope traits that implement
+the trait and define a field with that name. Sometimes it may be
+necessary or desirable to specify the trait explicitly. For those
+cases, we introduce a fully qualified field notation which looks like
+`x.<Trait<U,V>::f>`.
+
+This is comparable to the associated item notation `<T as
+Trait<U,V>>::foo`, but with some differences. First, the `Self` type
+is implied by the type of the left-hand side (`x`, in our
+example). Second, it is part of an lvalue expression, rather than
+being a path to an item.
+
+This is obviously a point where one might bikeshed on the best syntax.
+Here are some alternatives that we considered and rejected:
+
+- `x.Trait::f` more closely resembles C++ and Java and could work, but
+  it would have to be `x.Trait::<U,V>::f` if you wish to specify the trait's
+  type parameters.
+- `<x as Trait<U,V>>.f` doesn't work because we'd need a cover grammar for types
+  and expressions, since you cannot do not know whether `x` is a type or
+  expression until you see the `.`.
+- `x::<as Trait<U,V>>.f` seems random
+
+#### Leaving room for method calls
+
+It has not escaped our notice that this same syntax could be used for
+method calls. This RFC does not propose that we permit that, but we do
+restrict the grammar of call expressions to disallow a fully qualified
+field path on the left-hand-side of `()`. In other words, just as
+calling a closure located at `a.b.c` must be written `(a.b.c)()` (so
+as to avoid ambiguity with calling a method `c` on the path `a.b`), so
+must you wrote `(a.b.<Trait::c>)()`; `a.b.<Trait::c>()` will not
```

Comment:
To be clear, the paren restriction could be removed, as no trait can have a field and method with the same name, right?